src/shell/signal.c

```diff
@@ -0,0 +1,257 @@
+/************************************************************\
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* Send a signal to a job at a predefined value for timeleft
+ *
+ * Set via attributes.system.shell.options.signal, e.g.:
+ *  { "timeleft": 123, "signal": 10 }
+ *
+ * with defaults:
+ *  timeleft = 60.
+ *  signal = 10
+ *
+ * If shell.options.signal is not set in jobspec, then no warning
+ * signal is sent.
+ */
+#define FLUX_SHELL_PLUGIN_NAME "signal"
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <signal.h>
+#include <string.h>
+#include <jansson.h>
+#include <flux/core.h>
+#include <flux/shell.h>
+
+#include "src/common/libutil/fsd.h"
+#include "src/common/libutil/sigutil.h"
+#include "internal.h"
+#include "builtins.h"
+
+struct shell_signal {
+    flux_shell_t *shell;
+    flux_jobid_t id;
+    flux_watcher_t *watcher;
+    double timeleft;
+    int signum;
+};
+
+static void shell_signal_destroy (struct shell_signal *sig)
+{
+    if (sig) {
+        int saved_errno = errno;
+        flux_watcher_destroy (sig->watcher);
+        free (sig);
+        errno = saved_errno;
+    }
+}
+
+static void kill_cb (flux_future_t *f, void *arg)
+{
+    if (flux_future_get (f, NULL) < 0)
+        shell_log_error ("flux_job_kill");
+    flux_future_destroy (f);
+}
+
+static void shell_signal_cb (flux_reactor_t *r,
+                             flux_watcher_t *w,
+                             int revents,
+                             void *arg)
+{
+    struct shell_signal *sig = arg;
+    flux_future_t *f;
+
+    shell_log ("job will expire in %.1fs, sending %s to job",
+               sig->timeleft,
+               sigutil_signame (sig->signum));
+
+    if (!(f = flux_job_kill (sig->shell->h, sig->id, sig->signum))
+        || flux_future_then (f, -1., kill_cb, sig) < 0)
+        shell_log_error ("failed to send %s to job",
+                         sigutil_signame (sig->signum));
+}
+
+static int set_timeleft_watcher (struct shell_signal *sig)
+{
+    double expiration;
+    double wakeup;
+    double remaining;
+
+    if (flux_shell_info_unpack (sig->shell,
+                                "{s:{s:{s:f}}}",
+                                "R",
+                                "execution",
+                                "expiration", &expiration) < 0) {
+        shell_log_errno ("unable to get job expiration");
+        return -1;
+    }
+
+    /* Destroy any current watcher in case this function is called due
+     * to an update in a job's expiration (future update).
+     */
+    flux_watcher_destroy (sig->watcher);
+    sig->watcher = NULL;
+
+    if (expiration == 0.)
+        return 0;
```

Comment:
If the user specified `--signal` on a job with no expiration, should something be logged like
> job has no expiration time, SIG will not be sent

Reply:
Great idea, I'll add that.